File: DotNet/PubSubByType/Subscriber.cs
using System;
using System.Reactive;
using System.Reactive.Linq;
using Shastra.Hydra.Messaging;
using Shastra.Hydra.Messaging.Listeners;
using Shastra.Hydra.Messaging.MessageFetchers;
using Shastra.Hydra.Messaging.Serializers;

namespace Shastra.Hydra.PubSubByType
{
    /// <summary>
    /// Class for subscribing to Hydra messages of a given type. To be used in conjunction with the Publisher class.
    /// </summary>
    /// <typeparam name="TSub">The type of messages to subscribe to.</typeparam>
    public class Subscriber<TSub> : IObservable<Notification<AugmentedMessage<TSub>>>, IDisposable
    {
        private readonly IListener<HydraMessage> _listener;
        private readonly IObservable<Notification<AugmentedMessage<TSub>>> _messageSource;
        private readonly ISerializer<TSub> _serializer;

        public event Action<object, Notification<AugmentedMessage<TSub>>> MessageInQueue;

        public long BufferDelayMs { get { return _listener.BufferDelayMs; } set { _listener.BufferDelayMs = value; } }
        public long PollIntervalMs { get { return _listener.PollIntervalMs; } set { _listener.PollIntervalMs = value; } }

        /// <summary>
        /// Subscribe to all published messages of type TSub.
        /// </summary>
        /// <param name="hydraService">The Hydra service to use for sending messages.</param>
        /// <param name="serializer">The serializer to use for TSub objects. Defaults to HydraDataContractSerializer.</param>
        /// <param name="topic">The topic to listen for. Defaults to the full name of TSub.</param>
        /// <remarks>The serializer must match the one used by the message publisher.</remarks>
        public Subscriber(IHydraService hydraService, ISerializer<TSub> serializer = null, string topic = null) : this(hydraService, null, serializer, topic) {}

        /// <summary>
        /// Subscribe to all messages of type TSub sent to a specific destination.
        /// </summary>
        /// <param name="hydraService">The Hydra service to use for sending messages.</param>
        /// <param name="thisParty">The message destination (i.e. this app).</param>
        /// <param name="serializer">The serializer to use for TSub objects. Defaults to HydraDataContractSerializer.</param>
        /// <param name="topic">The topic to listen for. Defaults to the full name of TSub.</param>
        /// <remarks>The serializer must match the one used by the message publisher. If thisParty is null, then the subscription will be to all published TSub messages
        /// regardless of destination.</remarks>
        public Subscriber(IHydraService hydraService, string thisParty, ISerializer<TSub> serializer = null, string topic = null)
        {
            _serializer = serializer ?? new HydraDataContractSerializer<TSub>();
            string messageTopic = topic ?? typeof (TSub).FullName;
            if (thisParty == null) {
                _listener = hydraService.GetListener(new HydraByTopicMessageFetcher(messageTopic));
            } else {
                _listener = hydraService.GetListener(new HydraByTopicByDestinationMessageFetcher(messageTopic, thisParty));
            }
            _messageSource = _listener.Select(MessageNotification);
            _messageSource.Subscribe(MessageSourceOnNext);
        }

        private Notification<AugmentedMessage<TSub>> MessageNotification(HydraMessage message)
        {
            try {
                return Notification.CreateOnNext(new AugmentedMessage<TSub>(_serializer.Deserialize(message.Data), message.Attachments));
            } catch (Exception ex) {
                return Notification.CreateOnError<AugmentedMessage<TSub>>(ex);
            }
        }

        private void MessageSourceOnNext(Notification<AugmentedMessage<TSub>> message)
        {
            if (MessageInQueue != null) {
                MessageInQueue(this, message);
            }
        }

        #region Implementation of IObservable<out Notification<AugmentedMessage<TSub>>>

        public IDisposable Subscribe(IObserver<Notification<AugmentedMessage<TSub>>> observer)
        {
            return _messageSource.Subscribe(observer);
        }

        #endregion

        #region Implementation of IDisposable

        // See http://msdn.microsoft.com/en-us/library/ms244737.aspx

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (disposing) {
                // free managed resources
                _listener.Dispose();
            }
            // free native resources if there are any.
        }

        #endregion

    }
}

File: Java/hydra/src/main/java/uk/co/shastra/hydra/pubsubbytype/Subscriber.java
package uk.co.shastra.hydra.pubsubbytype;

import com.fasterxml.jackson.core.type.TypeReference;

import rx.Notification;
import rx.Observable;
import rx.util.functions.Action1;
import rx.util.functions.Func1;
import uk.co.shastra.hydra.messaging.HydraMessage;
import uk.co.shastra.hydra.messaging.HydraService;
import uk.co.shastra.hydra.messaging.listeners.Listener;
import uk.co.shastra.hydra.messaging.messagefetchers.HydraByTopicByDestinationMessageFetcher;
import uk.co.shastra.hydra.messaging.messagefetchers.HydraByTopicMessageFetcher;
import uk.co.shastra.hydra.messaging.serializers.HydraJsonSerializer;
import uk.co.shastra.hydra.messaging.serializers.Serializer;
import uk.co.shastra.hydra.messaging.utils.Event;
import uk.co.shastra.hydra.messaging.utils.EventHandler;

public class Subscriber<TSub> {

    private Listener<HydraMessage> listener;
    private Observable<Notification<TSub>> messageSource;
    private Serializer<TSub> serializer;
	
    private Event<Notification<TSub>> messageInQueue = new Event<Notification<TSub>>();

    /**
     * @return The time in milliseconds to delay delivery so that out-of-sequence messages can be delivered in the right order.
     */
    public long getBufferDelayMs() { return listener.getBufferDelayMs(); }
    /**
     * @param bufferDelayMs The time in milliseconds to delay delivery so that out-of-sequence messages can be delivered in the right order.
     */
    public void setBufferDelayMs(long bufferDelayMs) { listener.setBufferDelayMs(bufferDelayMs); }
    
    /**
     * @return Time between polls for messages, in milliseconds
     */
    public long getPollIntervalMs() { return listener.getPollIntervalMs(); }
    /**
     * @param pollIntervalMs Time between polls for messages, in milliseconds
     */
    public void setPollIntervalMs(long pollIntervalMs) { listener.setPollIntervalMs(pollIntervalMs); }
    
    /**
     * Subscribe to all published messages of type TSub
     * 
     * @param hydraService The Hydra service to use for sending messages
     * @param valueType The type into which incoming messages are deserialised. Usually TSub.class where TSub is the concrete value of this class's generic parameter
     */
    public Subscriber(HydraService hydraService, Class<TSub> valueType) { this(hydraService, valueType, null, null, null); }
    /**
     * Subscribe to all published messages of type TSub
     * 
     * The serializer must match the one used by the message publisher
     * 
     * @param hydraService The Hydra service to use for sending messages
     * @param valueType The type into which incoming messages are deserialised. Usually TSub.class where TSub is the concrete value of this class's generic parameter
     * @param serializer The serializer to use for TSub objects. Defaults to HydraJsonSerializer
     */
    public Subscriber(HydraService hydraService, Class<TSub> valueType, Serializer<TSub> serializer) { this(hydraService, valueType, null, serializer, null); }
    /**
     * Subscribe to all messages of type TSub sent to a specific destination
     * 
     * If thisParty is null, then the subscription will be to all published TSub messages regardless of destination
     * 
     * @param hydraService The Hydra service to use for sending messages
     * @param valueType The type into which incoming messages are deserialised. Usually TSub.class where TSub is the concrete value of this class's generic parameter
     * @param thisParty The message destination (i.e. this app)
     */
    public Subscriber(HydraService hydraService, Class<TSub> valueType, String thisParty) { this(hydraService, valueType, thisParty, null, null); }   
    /**
     * Subscribe to all messages of type TSub sent to a specific destination
     * 
     * The serializer must match the one used by the message publisher. If thisParty is null, then the subscription will be to all published TSub messages
     * regardless of destination
     * 
     * @param hydraService The Hydra service to use for sending messages
     * @param valueType The type into which incoming messages are deserialised. Usually TSub.class where TSub is the concrete value of this class's generic parameter.
     * @param thisParty The message destination (i.e. this app)
     * @param serializer The serializer to use for TSub objects. Defaults to HydraJsonSerializer
     * @param topic The topic to listen for. Defaults to the canonical name of valueType
     */
    public Subscriber(HydraService hydraService, Class<TSub> valueType, String thisParty, Serializer<TSub> serializer, String topic)
    {
        this.serializer = serializer == null ? new HydraJsonSerializer<TSub>(valueType) : serializer;
        String messageTopic = topic == null ? valueType.getCanonicalName() : topic;
        init(hydraService, thisParty, messageTopic);
    } 
    /**
     * Subscribe to all messages of type TSub
     * 
     * @param hydraService The Hydra service to use for sending messages
     * @param valueTypeRef The type into which incoming messages are deserialised. Usually new TypeReference<TSub>() {} and used when TSub is itself a
     * type with a generic parameter so that TSub.class cannot be used
     * @param topic The topic to listen for
     */
    public Subscriber(HydraService hydraService, TypeReference<?> valueTypeRef, String topic) { this(hydraService, valueTypeRef, topic, null); }
    /**
     * Subscribe to all messages of type TSub sent to a specific destination
     * 
     * @param hydraService The Hydra service to use for sending messages
     * @param valueTypeRef The type into which incoming messages are deserialised. Usually new TypeReference<TSub>() {} and used when TSub is itself a
     * type with a generic parameter so that TSub.class cannot be used
     * @param topic The topic to listen for
     * @param thisParty The message destination (i.e. this app)
     */
    public Subscriber(HydraService hydraService, TypeReference<?> valueTypeRef, String topic, String thisParty) { 
    	this(hydraService, new HydraJsonSerializer<TSub>(valueTypeRef), topic, thisParty);
    }
    /**
     * Subscribe to all messages of type TSub sent to a specific destination
     * 
     * @param hydraService The Hydra service to use for sending messages
     * @param serializer The serializer to use for TSub objects
     * @param topic The topic to listen for
     */
    public Subscriber(HydraService hydraService, Serializer<TSub> serializer, String topic) { this(hydraService, serializer, topic, null); }
    /**
     * Subscribe to all messages of type TSub sent to a specific destination
     * 
     * @param hydraService The Hydra service to use for sending messages
     * @param serializer The serializer to use for TSub objects
     * @param topic The topic to listen for
     * @param thisParty The message destination (i.e. this app)
     */ 
    public Subscriber(HydraService hydraService, Serializer<TSub> serializer, String topic, String thisParty) {
    	this.serializer = serializer;
    	init(hydraService, thisParty, topic);
    }
    
    private void init(HydraService hydraService, String thisParty, String topic) {
		if (thisParty == null) {
            listener = hydraService.getListener(new HydraByTopicMessageFetcher(topic));
        } else {
            listener = hydraService.getListener(new HydraByTopicByDestinationMessageFetcher(topic, thisParty));
        }
        messageSource = listener.getObservable().map(new Func1<HydraMessage, Notification<TSub>>() {
			@Override public Notification<TSub> call(HydraMessage message) {
				return MessageNotification(message.getData());
			}
		});
        messageSource.subscribe(new Action1<Notification<TSub>>() {
			@Override public void call(Notification<TSub> message) { messageInQueue.raise(this, message);	}
		});
	}

    private Notification<TSub> MessageNotification(String data)
    {
        try {
            return new Notification<TSub>(serializer.deserialize(data));
        } catch (Exception ex) {
            return new Notification<TSub>(ex);
        }
    }
    
    public void addMessageInQueueHandler(EventHandler<Notification<TSub>> handler) { messageInQueue.addHandler(handler); }
    public void removeMessageInQueueHandler(EventHandler<Notification<TSub>> handler) { messageInQueue.removeHandler(handler); }
    
    public Observable<Notification<TSub>> getObservable() { return messageSource; }
    
    public void close()
    {
    	listener.close();
    }
}
